lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1413,6 +1413,48 @@ RequirementSignatureRequest::evaluate(Evaluator &evaluator, ProtocolDecl *proto)
   return reqSignature->getRequirements();
 }
 
+llvm::Expected<Type>
+DefaultDefinitionTypeRequest::evaluate(Evaluator &evaluator,
+                                       AssociatedTypeDecl *assocType) const {
+  if (assocType->Resolver) {
+    auto defaultType = assocType->Resolver->loadAssociatedTypeDefault(
+                                    assocType, assocType->ResolverContextData);
+    assocType->Resolver = nullptr;
+    assocType->DefaultDefinition.setType(defaultType);
+    return defaultType;
+  }
+
+  auto &context = assocType->getASTContext();
+  auto &TC = TypeChecker::createForContext(assocType->getASTContext());
+
+  TypeLoc &defaultDefinition = assocType->DefaultDefinition;
+  if (!defaultDefinition.isNull()) {
+    if (TC.validateType(defaultDefinition,
+                        TypeResolution::forInterface(
+                                                  assocType->getDeclContext()),
+                        None)) {
+      defaultDefinition.setInvalidType(context);
+    } else {
+      // associatedtype X = X is invalid
+      auto mentionsItself =
+        defaultDefinition.getType().findIf([&](Type type) {
+          if (auto DMT = type->getAs<DependentMemberType>()) {
+            return DMT->getAssocType() == assocType;
+          }
+          return false;
+        });
+
+      if (mentionsItself) {
+        TC.diagnose(defaultDefinition.getLoc(), diag::recursive_decl_reference,
```

Comment:
assocType->diagnose(...)

Reply:
After moving this check to DeclChecker::visitAssociatedTypeDecl where there is access to a TypeChecker, I left this call on TC to keep the diagnose arrow pointing at the default definition.